pkg/render/goldmane/component_test.go

```diff
@@ -0,0 +1,135 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package goldmane_test
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+
+	operatorv1 "github.com/tigera/operator/api/v1"
+	"github.com/tigera/operator/pkg/ptr"
+	"github.com/tigera/operator/pkg/render"
+	rmeta "github.com/tigera/operator/pkg/render/common/meta"
+	"github.com/tigera/operator/pkg/render/common/securitycontext"
+	rtest "github.com/tigera/operator/pkg/render/common/test"
+	"github.com/tigera/operator/pkg/render/goldmane"
+	"github.com/tigera/operator/pkg/tls/certificatemanagement"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	defaultTLSKeyPair        = certificatemanagement.NewKeyPair(&corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: "key-pair"}}, nil, "")
+	defaultTrustedCertBundle = certificatemanagement.CreateTrustedBundle(nil)
+)
+
+var _ = Describe("ComponentRendering", func() {
+	DescribeTable("Creation and deletion counts", func(cfg *goldmane.Configuration, creatObjs, delObjs int) {
+		component := goldmane.Goldmane(cfg)
+		objsToCreate, objsToDelete := component.Objects()
+		Expect(objsToCreate).To(HaveLen(creatObjs))
+		Expect(objsToDelete).To(HaveLen(delObjs))
+	},
+		Entry("Should return objects to create when variant is Calico",
+			&goldmane.Configuration{
+				Installation: &operatorv1.InstallationSpec{
+					KubernetesProvider: operatorv1.ProviderGKE,
+					Variant:            operatorv1.Calico,
+				},
+				TrustedCertBundle:     certificatemanagement.CreateTrustedBundle(nil),
+				GoldmaneServerKeyPair: defaultTLSKeyPair,
+			},
+			6, 0,
+		),
+		Entry("Should return objects to delete when variant is not Calico",
+			&goldmane.Configuration{
+				Installation: &operatorv1.InstallationSpec{
+					KubernetesProvider: operatorv1.ProviderGKE,
+					Variant:            operatorv1.TigeraSecureEnterprise,
+				},
+				TrustedCertBundle:     certificatemanagement.CreateTrustedBundle(nil),
+				GoldmaneServerKeyPair: defaultTLSKeyPair,
+			},
+			0, 6,
+		),
+	)
+
+	DescribeTable("Goldmane Deployment", func(cfg *goldmane.Configuration, expected *appsv1.Deployment) {
+		component := goldmane.Goldmane(cfg)
+		objsToCreate, _ := component.Objects()
+
+		deployment, err := rtest.GetResourceOfType[*appsv1.Deployment](objsToCreate, goldmane.GoldmaneName, goldmane.GoldmaneNamespace)
+		Expect(err).ShouldNot(HaveOccurred())
+		Expect(deployment).To(Equal(expected))
```

Comment:
A bit of a nit, but the existing tests you've written for Whisker for example are exceptionally tricky to debug when they fail because the error message is so huge - do you have some tip for how to do that?

It would be really nice if we could factor these in a way where the precise diff was presented in the error message, and if we could do that before getting too deep into this testing pattern (most of our other tests check fields individually rather than a full equality on two fully specified Deployments)

Reply:
Ya the way I do it is add the cmp.Diff function, and add the output to the comment. Maybe I should just keep those cmp.Diff calls in here and commit it.

Reply:
yeah, it's a no-no if I need to edit the test in order to understand its output. Why not just commit the good stuff?

Reply:
I've just been doing it way to long to tests I don't want to commit changes for and it's become a bad habit to not commit it lol.

I'll commit it.

Reply:
done